### PR TITLE
Splitting Number into int and float

### PR DESCRIPTION
**Pull request by Lythenas**
Int and float are the lua terms. In C++ this is implemented as long and double.

TODO:

- [ ] Fix parsing
- [ ] Figure out why some of the operators broke (`#` and `<`)

### CODE REVIEW

**Review by tn4799 — CHANGES_REQUESTED** [review]
If I written some functions wrong, that was not intended. But I think my thoughs should be clear what I want to achieve.

File: include/MiniLua/values.hpp [hunk 1]
```diff
@@ -822,6 +897,7 @@ class Value {
     Value(bool val);
     Value(Number val);
     Value(int val);
```
[the remark this answers is not in the document]
Comment (Lythenas) [reply include/MiniLua/values.hpp:899]: I think C++ does not coerce an int to a long. At least it complains about not finding a constructor if you try to call it with an int.

File: src/parse_literals.cpp [hunk 2]
```diff
@@ -1,25 +1,60 @@
...
 auto parse_number_literal(const std::string& str) -> Value {
-    std::regex pattern_number(R"(\s*\d+\.?\d*)");
-    std::regex pattern_hex(R"(\s*0[xX][\dA-Fa-f]+\.?[\dA-Fa-f]*)");
-    std::regex pattern_exp(R"(\s*\d+\.?\d*[eE]-?\d+)");
+    // if the string matches the expected format we parse it otherwise return nil
+    if (std::regex_match(str, to_number_int_pattern)) {
+        // parse as int and check if whole input is consumed
+        try {
+            char* last_pos;
+            // we need to parse as unsigned long to retain the same behaviour as lua
```
[the remark this answers is not in the document]
Comment (Lythenas) [reply src/parse_literals.cpp:24]: Yes but if you parse e.g. `0xfff...fff` it should be -1. If we parse it as a signed long it will just overflow (or be the max long, I don't remember). If we parse it as an unsigned long and the cast it works the same as in lua. And the parse functions for unsigned longs also accepts `-something`. I have to write a few more tests to confirm that this works in all cases (especially for literals that are out of range) but all current tests pass.

File: tests/public_api/values.cpp [hunk 3]
```diff
@@ -101,7 +101,8 @@ TEST_CASE("number Value is constructable") {
         const minilua::Value value{2};
         CHECK(std::holds_alternative<minilua::Number>(value.raw()));
         CHECK(std::get<minilua::Number>(value.raw()) == 2);
-        CHECK(std::get<minilua::Number>(value.raw()).value == 2);
+        CHECK(std::get<minilua::Number>(value.raw()).try_as_int() == 2);
+        CHECK(std::get<minilua::Number>(value.raw()).as_float() == 2);
```
[the remark this answers is not in the document]
Comment (Lythenas) [reply tests/public_api/values.cpp:105]: Types are automatically converted to doubles. But this is maybe a bit clearer.

File: src/math.cpp [hunk 4]
```diff
@@ -759,25 +790,34 @@ auto ult(const CallContext& ctx) -> Value {
...
+            if (m.is_int()) {
+                m_int = m.try_as_int();
+            } else {
+                double num_m;
+                double fraction = std::modf(m.as_float(), &num_m);
+                if (fraction != 0.0) {
+                    throw std::runtime_error(
+                        "bad argument #1 to 'ult' (number has no integer representation)");
+                }
+                m_int = (int)num_m;
```
[the remark this answers is not in the document]
Comment (Lythenas) [reply src/math.cpp:804]: This is not correct. Because you can call ult with float values (but they have to be whole numbers). That's why it says "no integer **representation**".

File: src/math.cpp [hunk 5]
```diff
@@ -759,25 +790,34 @@ auto ult(const CallContext& ctx) -> Value {
...
+            } else {
                 double num_n;
-                fraction = std::modf(n, &num_n);
-                if (fraction == 0.0) {
-                    unsigned long ml = (long)num_m;
-                    unsigned long nl = (long)n;
-                    return ml < nl;
-                } else {
+                double fraction = std::modf(n.as_float(), &num_n);
+                if (fraction != 0.0) {
```
[the remark this answers is not in the document]
Comment (Lythenas) [reply src/math.cpp:813]: Same as above

File: src/math.cpp [hunk 6]
```diff
@@ -759,25 +790,34 @@ auto ult(const CallContext& ctx) -> Value {
...
                     throw std::runtime_error(
                         "bad argument #2 to 'ult' (number has no integer representation)");
                 }
-            } else {
-                throw std::runtime_error(
-                    "bad argument #1 to 'ult' (number has no integer representation)");
+                n_int = (int)num_n;
```
[the remark this answers is not in the document]
Comment (Lythenas) [reply src/math.cpp:817]: Same as above. Also not that `try_as_int` will throw an exception if the value is not an int it does not check if it is a whole number.

File: src/parse_literals.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lythenas) [reply src/parse_literals.cpp:24]: As it turns out this will not parse correctly if the value is out of range. But I think none of the C/C++ functions will parse it correctly if it is out of range.

File: src/math.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tn4799) [reply src/math.cpp:804]: ok, then i did remember it wrong and the code is fine like that

**Review by tn4799 — APPROVED** [review]
